src/login/controller/login.ts

```diff
@@ -56,9 +56,23 @@ class LoginController extends Controller {
         throw err;
       }
     }
-    const user = await principalService.findByIdentity(identity) as User;
+
+    const user = (await principalService.findByIdentity(identity)) as User;
+
+    if (await services.loginActivity.isAccountLocked(user)) {
+      await services.loginActivity.incrementFailedLoginAttempts(user);
+      log(EventType.loginFailedAccountLocked, ctx.ip(), user.id, ctx.request.headers.get('User-Agent'));
+      return this.redirectToLogin(ctx, '', `Too many failed login attempts, please contact ${requireSetting('smtp.emailFrom')} to unlock your account.`);
+    }
 
     if (!await services.user.validatePassword(user, ctx.request.body.password)) {
```

Comment:
This is a bit of a larger request, but I wonder if _most_ of this logic should go into the service. The reason is that this is not the only place where we check a username and password, there are two more places:

* OAuth2 has a `password` flow
* I recently added a 'authorization_challenge` OAuth2 flow.

These 3 features all do a username/password check so it kinda makes sense that they all increment the 'failed login' counter if a password fails.

I'm not sure if this should just be part of 'validatePassword' or a new function.

Reply:
Updated: 33288d2629c953dbb59581a631739c57e04cf7d8
Could you double-check the use cases of OAuth2 password flow and the OAuth2 authorization_challenge flow?

Reply:
Gonna do this while working on #513